spec/active_record/connection_adapters/mysql2_ghost_adapter_spec.rb

```diff
@@ -25,57 +29,147 @@
+            subject.add_index(table, column, using: method)
+          end
+        end
+
+        context 'with type option' do
+          let(:type) { 'fulltext' }
+
+          it 'passes the correct SQL to #execute' do
+            expect_sql "ALTER TABLE `#{table}` ADD #{type.upcase} INDEX `index_#{table}_on_#{column}` (`#{column}`)"
+            subject.add_index(table, column, type: type)
+          end
+        end
       end
 
-      it 'passes the built SQL to #execute' do
-        expect(subject).to(
-          receive(:execute)
-          .with("ALTER TABLE `#{table_name}` DROP INDEX `index_name`")
-        )
-        subject.remove_index(table_name, options)
+      describe '#remove_index' do
+        let(:index_name) { 'index_name' }
+
+        before do
+          # This must be mocked because the method tries to query the database to get the index name
+          allow(subject).to receive(:index_name_for_remove).and_return(index_name)
+        end
+
+        context 'with column name passed as positional argument' do
+          it 'passes the correct SQL to #execute' do
+            expect_sql "ALTER TABLE `#{table}` DROP INDEX `#{index_name}`"
+            subject.remove_index(table, column)
+          end
+        end
+
+        context 'with column name passed as keyword argument' do
+          it 'passes the correct SQL to #execute' do
+            expect_sql "ALTER TABLE `#{table}` DROP INDEX `#{index_name}`"
+            subject.remove_index(table, column: column)
+          end
+        end
+
+        context 'with index name passed as keyword argument' do
+          let(:index_name) { 'brand_new_index_name' }
+
+          it 'passes the correct SQL to #execute' do
+            expect_sql "ALTER TABLE `#{table}` DROP INDEX `#{index_name}`"
+            subject.remove_index(table, name: index_name)
+          end
+        end
+
+        if Gem.loaded_specs['activerecord'].version >= Gem::Version.new('6.1')
+          context 'when ActiveRecord version is >= 6.1' do
+            context 'with if_exists property set false' do
+              context 'when index exists' do
+                before { allow(subject).to receive(:index_exists?).and_return(true) }
+
+                it 'passes the correct SQL to #execute' do
+                  expect_sql "ALTER TABLE `#{table}` DROP INDEX `#{index_name}`"
+                  subject.remove_index(table, column, if_exists: true)
+                end
+              end
+
+              context 'when index does not exist' do
+                before { allow(subject).to receive(:index_exists?).and_return(false) }
+
+                it 'does nothing' do
+                  subject.remove_index(table, column, if_exists: true)
+                  expect(subject).not_to have_received(:execute)
+                end
+              end
+            end
+          end
+        end
       end
     end
   end
 end
+
+def expect_sql(sql)
+  expect(subject).to receive(:execute).with(sql)
```

Comment:
because there's a `SELECT` query that happens before each `ALTER TABLE` query, we can't use `have_received` (which rubocop says is better?)

Reply:
You can also make custom matchers through the `spec_helper`. https://relishapp.com/rspec/rspec-expectations/docs/custom-matchers/define-a-custom-matcher

Reply:
yeah good call; will update